binding.cc

```diff
@@ -2,189 +2,147 @@
-  SN_ARGV(1, sodium_mlock)
-
-  SN_ARGV_TYPEDARRAY(buf, 0)
-
-  SN_RETURN(sodium_mlock(buf_data, buf_size), "memory lock failed")
+static inline int
+sn_sodium_mlock(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> buf
+) {
+  return sodium_mlock(buf.data(), buf.size_bytes());
 }
 
-js_value_t *
-sn_sodium_munlock (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_munlock)
-
-  SN_ARGV_TYPEDARRAY(buf, 0)
-
-  SN_RETURN(sodium_munlock(buf_data, buf_size), "memory unlock failed")
+static inline int
+sn_sodium_munlock(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> buf
+) {
+  return sodium_munlock(buf.data(), buf.size_bytes());
 }
 
-static void sn_sodium_free_finalise (js_env_t *env, void *finalise_data, void *finalise_hint) {
+static void
+sn_external_arraybuffer_finalize(js_env_t *env, void *finalise_data, void *finalise_hint) {
   sodium_free(finalise_data);
-
-  int64_t ext_mem;
-  int err = js_adjust_external_memory(env, -4 * 4096, &ext_mem);
-  assert(err == 0);
 }
 
-js_value_t *
-sn_sodium_free (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_free)
-
-  SN_ARGV_TYPEDARRAY_PTR(buf, 0)
-  if (buf_data == NULL) return NULL;
-
-  js_value_t *array_buf;
-  SN_STATUS_THROWS(js_get_named_property(env, argv[0], "buffer", &array_buf), "failed to get arraybuffer");
+static js_arraybuffer_t
+sn_sodium_malloc(js_env_t *env, js_receiver_t, uint32_t len) {
+  void *ptr = sodium_malloc(len);
+  assert(ptr != nullptr);
 
-  SN_STATUS_THROWS(js_detach_arraybuffer(env, array_buf), "failed to detach array buffer");
-
-  void *ptr;
-  err = js_remove_wrap(env, array_buf, &ptr);
+  js_value_t *buffer;
+  int err = js_create_external_arraybuffer(env, ptr, len, sn_external_arraybuffer_finalize, NULL, &buffer);
   assert(err == 0);
-  assert(ptr == buf_data);
 
-  sodium_free(buf_data);
-
-  int64_t ext_mem;
-  err = js_adjust_external_memory(env, -4 * 4096, &ext_mem);
-  assert(err == 0);
-  return NULL;
+  return static_cast<js_arraybuffer_t>(buffer);
 }
 
-js_value_t *
-sn_sodium_malloc (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_malloc);
-
-  SN_ARGV_UINT32(size, 0)
-
-  void *ptr = sodium_malloc(size);
-  SN_THROWS(ptr == NULL, "ENOMEM")
-
-  SN_THROWS(ptr == NULL, "sodium_malloc failed");
-
-  js_value_t *buffer;
-
-  SN_STATUS_THROWS(js_create_external_arraybuffer(env, ptr, size, NULL, NULL, &buffer), "failed to create a native arraybuffer")
-
+static void
+sn_sodium_free(js_env_t *env, js_receiver_t, js_arraybuffer_t buf) {
+  int err;
 
-  js_value_t *value;
-  SN_STATUS_THROWS(js_get_boolean(env, true, &value), "failed to create boolean")
-  SN_STATUS_THROWS(js_set_named_property(env, buffer, "secure", value), "failed to set secure property")
+  js_value_t *handle = static_cast<js_value_t *>(buf);
 
-  err = js_wrap(env, buffer, ptr, sn_sodium_free_finalise, NULL, NULL);
+  bool is_detached;
+  err = js_is_detached_arraybuffer(env, handle, &is_detached);
```

Comment:
This check isn't necessary, the returned view will simply be empty when the buffer is detached.